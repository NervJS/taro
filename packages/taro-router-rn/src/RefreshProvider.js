import React from 'react'
import { ScrollView, RefreshControl } from 'react-native'

class RefreshProvider extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      refreshing: false
    }
  }

  render () {
    const {enablePullDownRefresh, disableScroll} = this.props
    if (disableScroll) {
      return this.props.children
    } else {
      return (
        <ScrollView
          style={{flex: 1}}
          scrollEventThrottle={5}
          alwaysBounceVertical={false}
          onScroll={this.onScroll}
          refreshControl={
            enablePullDownRefresh
              ? <RefreshControl
                refreshing={this.state.refreshing}
                onRefresh={this.handlePullDownRefresh}
              />
              : null
          }
        >
          {this.props.children && this.props.children}
        </ScrollView>
      )
    }
  }

  onScroll = (e) => {
    this.props.onScroll && this.props.onScroll(e.nativeEvent.contentOffset.y)
  }

  // TODO 滚动到底部事件
  handleReachBottom = () => {
    this.props.onReachBottom && this.props.onReachBottom()
  }

  handlePullDownRefresh = (callback) => {
    this.setState({refreshing: true}, callback)
    try {
      // TODO 处理异步的情况
      this.props.onPullDownRefresh && this.props.onPullDownRefresh()
    } catch (e) {
      throw new Error(e)
    } finally {
      this.setState({refreshing: false})
    }
  }

  stopPullDownRefresh = (callback) => {
    this.setState({refreshing: false}, callback)
  }
}

export default RefreshProvider
