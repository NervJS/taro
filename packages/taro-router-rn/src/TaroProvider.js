import React from 'react'
import queryString from 'query-string'
import RefreshProvider from './RefreshProvider'

class TaroProvider extends React.Component {
  constructor (props, context) {
    super(props, context)
    this.refreshProviderRef = React.createRef()
  }

  componentDidMount () {
    let {Taro} = this.props
    // didFocus
    this.didFocusSubscription = this.props.navigation.addListener(
      'didFocus',
      payload => {
        // 页面进入后回退并不会调用 React 生命周期，需要在路由生命周期中绑定 this
        Taro.navigateTo = this.wxNavigateTo.bind(this)
        Taro.redirectTo = this.wxRedirectTo.bind(this)
        Taro.navigateBack = this.wxNavigateBack.bind(this)
        Taro.switchTab = this.wxSwitchTab.bind(this)
        Taro.getCurrentPages = this.wxGetCurrentPages.bind(this)
      }
    )
    try {
      Taro.startPullDownRefresh = this.refreshProviderRef.current && this.refreshProviderRef.current.handlePullDownRefresh
      Taro.stopPullDownRefresh = this.refreshProviderRef.current && this.refreshProviderRef.current.stopPullDownRefresh
    } catch (e) {
      console.log('this.refreshProviderRef: ')
      console.log(this.refreshProviderRef)
      throw e
    }
  }

  wxNavigateTo (params) {
    if (typeof params !== 'object') {
      console.warn('Taro.NavigateTo 参数必须为 object')
      return
    }
    let {url, success, fail, complete} = params
    if (url.startsWith('/')) {
      url = url.substr(1)
    }

    let obj = queryString.parseUrl(url)
    console.log(obj)
    try {
      this.props.navigation.push(obj.url, obj.query)
    } catch (e) {
      fail && fail(e)
      complete && complete(e)
      throw e
    }
    success && success()
    complete && complete()
  }

  wxRedirectTo (params) {
    if (typeof params !== 'object') {
      console.warn('Taro.RedirectTo 参数必须为 object')
      return
    }
    let {url, success, fail, complete} = params
    if (url.startsWith('/')) {
      url = url.substr(1)
    }

    let obj = queryString.parseUrl(url)
    console.log(obj)
    try {
      this.props.navigation.replace(obj.url, obj.query)
    } catch (e) {
      fail && fail(e)
      complete && complete(e)
      throw e
    }
    success && success()
    complete && complete()
  }

  wxSwitchTab (params) {
    if (typeof params !== 'object') {
      console.warn('Taro.SwitchTab 参数必须为 object')
      return
    }
    let {url, success, fail, complete} = params
    if (url.startsWith('/')) {
      url = url.substr(1)
    }

    let obj = queryString.parseUrl(url)
    console.log(obj)
    try {
      this.props.navigation.navigate(obj.url, obj.query)
    } catch (e) {
      fail && fail(e)
      complete && complete(e)
      throw e
    }
    success && success()
    complete && complete()
  }

  wxNavigateBack (params) {
    if (typeof params !== 'object') {
      console.warn('Taro.NavigateBack 参数必须为 object')
      return
    }
    let {delta = 1} = params
    while (delta > 0) {
      this.props.navigation.goBack()
      delta--
    }
  }

  wxGetCurrentPages () {
    let parentState = this.props.navigation.dangerouslyGetParent().state
    if (parentState && parentState.routes) {
      return parentState.routes.map(item => item.routeName)
    } else {
      return []
    }
  }

  componentWillUnmount () {
    // Remove the listener when you are done
    this.didFocusSubscription && this.didFocusSubscription.remove()
  }

  render () {
    return (
      <RefreshProvider
        ref={this.refreshProviderRef}
        {...this.props}
      >
        {this.props.children}
      </RefreshProvider>
    )
  }
}

export default TaroProvider
