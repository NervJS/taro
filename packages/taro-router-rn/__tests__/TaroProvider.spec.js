import React from 'react'
import { View } from 'react-native'
import TaroProvider from '../src/TaroProvider'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import RefreshProvider from '../src/RefreshProvider'

class HomeScreen extends React.Component {
  constructor (props) {
    super(props)
  }

  render () {
    return <View>Test</View>
  }
}

describe('TaroProvider', function () {
  let Taro = {}
  it('should render success', function () {
    const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  /**
   * Taro.navigateTo
   * Taro.redirectTo
   * Taro.navigateBack
   * Taro.switchTab
   * Taro.getCurrentPages
   * Taro.startPullDownRefresh
   * Taro.stopPullDownRefresh
   */
  describe('Taro API', function () {
    it('should Taro API mount success', function () {
      shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      const expected = ['navigateTo', 'redirectTo', 'navigateBack', 'switchTab', 'getCurrentPages', 'startPullDownRefresh', 'stopPullDownRefresh']
      expect(Object.keys(Taro)).toEqual(expected)
    })

    it('should call navigateTo success', function () {
      const mockCallback = jest.fn()
      const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      wrapper.setProps({navigation: {push: mockCallback}})
      Taro.navigateTo({url: '/pages/index'})
      expect(mockCallback.mock.calls.length).toBe(1)
    })

    it('should call redirectTo success', function () {
      const mockCallback = jest.fn()
      const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      wrapper.setProps({navigation: {replace: mockCallback}})
      Taro.redirectTo({url: '/pages/index'})
      expect(mockCallback.mock.calls.length).toBe(1)
    })

    it('should call navigateBack success', function () {
      const mockCallback = jest.fn()
      const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      wrapper.setProps({navigation: {goBack: mockCallback}})
      Taro.navigateBack({url: '/pages/index'})
      expect(mockCallback.mock.calls.length).toBe(1)
    })

    it('should call switchTab success', function () {
      const mockCallback = jest.fn()
      const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      wrapper.setProps({navigation: {navigate: mockCallback}})
      Taro.switchTab({url: '/pages/index'})
      expect(mockCallback.mock.calls.length).toBe(1)
    })

    it('should call getCurrentPages success', function () {
      const mockCallback = jest.fn().mockReturnValue({state: {}})
      const wrapper = shallow(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      wrapper.setProps({navigation: {dangerouslyGetParent: mockCallback}})
      Taro.getCurrentPages({url: '/pages/index'})
      expect(mockCallback.mock.calls.length).toBe(1)
    })

    it('should has refreshProviderRef', function () {
      const wrapper = mount(<TaroProvider Taro={Taro}><HomeScreen /></TaroProvider>)
      expect(wrapper.instance().refreshProviderRef.current).toBeInstanceOf(RefreshProvider)
    })
  })
})
