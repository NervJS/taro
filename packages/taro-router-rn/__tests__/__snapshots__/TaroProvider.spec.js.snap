// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaroProvider should render success 1`] = `
<RefreshProvider
  Taro={
    Object {
      "getCurrentPages": [Function],
      "navigateBack": [Function],
      "navigateTo": [Function],
      "redirectTo": [Function],
      "startPullDownRefresh": null,
      "stopPullDownRefresh": null,
      "switchTab": [Function],
    }
  }
>
  <HomeScreen />
</RefreshProvider>
`;
